library/common/network/configurator.cc

```diff
@@ -167,13 +172,36 @@ void Configurator::reportNetworkUsage(envoy_netconf_t configuration_key, bool ne
 
   // If configuration state changed, refresh dns.
   if (configuration_updated) {
-    refreshDns(configuration_key);
+    refreshDns(configuration_key, false);
+  }
+}
+
+void Configurator::onDnsResolutionComplete(
+    const std::string& host, const Extensions::Common::DynamicForwardProxy::DnsHostInfoSharedPtr&,
+    Network::DnsResolver::ResolutionStatus status) {
+  if (enable_drain_post_dns_refresh_ && pending_drain_) {
+    pending_drain_ = false;
+    if (status == Network::DnsResolver::ResolutionStatus::Success) {
+      ENVOY_LOG_EVENT(debug, "netconf_post_dns_drain_cx", host);
+      cluster_manager_.drainConnections(nullptr);
+    }
+  }
+}
+
+void Configurator::setDrainPostDnsRefreshEnabled(bool enabled) {
+  enable_drain_post_dns_refresh_ = enabled;
+  if (!enabled) {
+    pending_drain_ = false;
+  } else if (!dns_callbacks_handle_) {
+    if (auto dns_cache = dnsCache()) {
+      dns_callbacks_handle_ = dns_cache->addUpdateCallbacks(*this);
```

Comment:
For simplicity I think it would be simpler to just always have the callbacks registered but up to you.

Reply:
Added clarifying comment to the code here.